src/vs/code/node/cli.ts

```diff
@@ -319,23 +319,89 @@ export async function main(argv: string[]): Promise<any> {
 			options['stdio'] = 'ignore';
 		}
 
-		const child = spawn(process.execPath, argv.slice(2), options);
+		let child: ChildProcess;
+		if (platform() !== 'darwin') {
+			child = spawn(process.execPath, argv.slice(2), options);
 
-		if (args.wait && waitMarkerFilePath) {
-			return new Promise<void>(resolve => {
+			if (args.wait && waitMarkerFilePath) {
+				return new Promise<void>(resolve => {
 
-				// Complete when process exits
-				child.once('exit', () => resolve(undefined));
+					// Complete when process exits
+					child.once('exit', () => resolve(undefined));
 
-				// Complete when wait marker file is deleted
-				whenDeleted(waitMarkerFilePath!).then(resolve, resolve);
-			}).then(() => {
+					// Complete when wait marker file is deleted
+					whenDeleted(waitMarkerFilePath!).then(resolve, resolve);
+				}).then(() => {
 
-				// Make sure to delete the tmp stdin file if we have any
-				if (stdinFilePath) {
-					unlinkSync(stdinFilePath);
+					// Make sure to delete the tmp stdin file if we have any
+					if (stdinFilePath) {
+						unlinkSync(stdinFilePath);
+					}
+				});
+			}
+		} else {
+			const requiresWait = args['status'] || args['verbose'] || hasReadStdinArg;
+			const envVars: string[] = [];
+			for (const envKey in env) {
+				if (envKey === '_') {
+					continue;
 				}
-			});
+				const value = env[envKey];
+				envVars.push('--env');
+				envVars.push(envKey + '=' + value ?? '');
+			}
+			const openArgs: string[] = ['-n'];
+
+			let tmpfile = '';
+			let tmpfileError = '';
+			if (requiresWait) {
+				const time = new Date().getTime();
+				tmpfile = `${tmpdir()}/vscode-wait-transfer-${time}.log`;
+				tmpfileError = `${tmpdir()}/vscode-wait-transfer-error-${time}.log`;
+				writeFileSync(tmpfile, '');
+				openArgs.push('-W');
+				openArgs.push('--stdout', tmpfile);
+				openArgs.push('--stderr', tmpfileError);
+				argv.push('--wait');
+			}
+			const argsArr: string[] = [];
+			const execPathToUse = args['exec-path'] ?? process.execPath;
+			argsArr.push('-a', execPathToUse);
+			argsArr.push(...envVars, ...openArgs, '--args', ...argv.slice(2));
+			child = spawn('open', argsArr, options);
+
+			if (requiresWait) {
+				const openPromise = (child: ChildProcess) => new Promise<void>((c) => {
+					console.log('Temp file location: ' + tmpfile);
+					if (args['verbose']) {
+						const stream = openSync(tmpfile, 'r');
```

Comment:
Does this have to be sync? 

I wonder whether it makes sense to encapsulate something reusable here to keep the code lean, i.e. a polling file watcher that provides access to the content, similar to what we have here:

https://github.com/microsoft/vscode/blob/90347144857be0a9977b6636e3680718512e67a2/src/vs/workbench/contrib/output/common/outputChannelModel.ts#L143-L143

Reply:
I wrote a new file `cliVerboseLogger.ts` which provides a much more async filewatcher-based impl